Use hash func to boost file creation and lookup

RoyWFHuang opened the pull request:
Previously, SimpleFS used a sequential insertion method to create files, which worked efficiently when the filesystem contained only a small number of files.
However, in real-world use cases, filesystems often manage a large number of files, making sequential search and insertion inefficient.
Inspired by Ext4’s hash-based directory indexing, this change adopts a hash function to accelerate file indexing and improve scalability.

Change:
Implemented hash-based file index lookup
Improved scalability for large directory structures

hash_code = file_hash(file_name);

extent index = hash_code / SIMPLEFS_MAX_BLOCKS_PER_EXTENT
block index = hash_code % SIMPLEFS_MAX_BLOCKS_PER_EXTENT;
```
 inode
  +-----------------------+
  | i_mode = IFDIR | 0755 |      block 123 (simplefs_file_ei_block)
  | ei_block = 123    ----|--->  +----------------+
  | i_size = 4 KiB        |      | nr_files  = 7  |
  | i_blocks = 1          |      |----------------|
  +-----------------------+    0 | ee_block  = 0  |
              (extent index = 0) | ee_len    = 8  |
                                 | ee_start  = 84 |--->  +-------------+ block 84(simplefs_dir_block)
                                 | nr_file   = 2  |      |nr_files = 2 | (block index = 0)
                                 |----------------|      |-------------|
                               1 | ee_block  = 8  |    0 | inode  = 24 |
              (extent index = 1) | ee_len    = 8  |      | nr_blk = 1  |
                                 | ee_start  = 16 |      | (foo)       |
                                 | nr_file   = 5  |      |-------------|
                                 |----------------|    1 | inode  = 45 |
                                 | ...            |      | nr_blk = 14 |
                                 |----------------|      | (bar)       |
                             341 | ee_block  = 0  |      |-------------|
            (extent index = 341) | ee_len    = 0  |      | ...         |
                                 | ee_start  = 0  |      |-------------|
                                 | nr_file   = 12 |   14 | 0           |
                                 +----------------+      +-------------+ block 85(simplefs_dir_block)
                                                         |nr_files = 2 | (block index = 1)
                                                         |-------------|
                                                       0 | inode  = 48 |
                                                         | nr_blk = 15 |
                                                         | (foo1)      |
                                                         |-------------|
                                                       1 | inode  = 0  |
                                                         | nr_blk = 0  |
                                                         |             |
                                                         |-------------|
                                                         | ...         |
                                                         |-------------|
                                                      14 | 0           |
                                                         +-------------+
```
Performance test

* Random create 30600 files into filesystem

legacy:
```
         168140.12 msec task-clock                       #    0.647 CPUs utilized
            111367      context-switches                 #  662.346 /sec
             40917      cpu-migrations                   #  243.351 /sec
           3736053      page-faults                      #   22.220 K/sec
      369091680702      cycles                           #    2.195 GHz
      168751830643      instructions                     #    0.46  insn per cycle
       34044524391      branches                         #  202.477 M/sec
         768151711      branch-misses                    #    2.26% of all branches

     259.842753513 seconds time elapsed
      23.000247000 seconds user
     150.380145000 seconds sys
```

full_name_hash
```
         167926.13 msec task-clock                       #    0.755 CPUs utilized
            110631      context-switches                 #  658.808 /sec
             43835      cpu-migrations                   #  261.037 /sec
           3858617      page-faults                      #   22.978 K/sec
      392878398961      cycles                           #    2.340 GHz
      207287412692      instructions                     #    0.53  insn per cycle
       42556269864      branches                         #  253.423 M/sec
         840868990      branch-misses                    #    1.98% of all branches

     222.274028604 seconds time elapsed
      20.794966000 seconds user
     151.941876000 seconds sys
```

--------
* Random remove 30600 files into filesystem

legacy:
```
         104332.44 msec task-clock                       #    0.976 CPUs utilized
             56514      context-switches                 #  541.672 /sec
              1174      cpu-migrations                   #   11.252 /sec
           3796962      page-faults                      #   36.393 K/sec
      258293481279      cycles                           #    2.476 GHz
      153853176926      instructions                     #    0.60  insn per cycle
       30434271757      branches                         #  291.705 M/sec
         532967347      branch-misses                    #    1.75% of all branches

     106.921706288 seconds time elapsed
      16.987883000 seconds user
      91.268661000 seconds sys
```

full_name_hash
```
          83278.61 msec task-clock                       #    0.967 CPUs utilized
             52431      context-switches                 #  629.585 /sec
              1309      cpu-migrations                   #   15.718 /sec
           3796501      page-faults                      #   45.588 K/sec
      199894058328      cycles                           #    2.400 GHz
      110625460371      instructions                     #    0.55  insn per cycle
       20325767251      branches                         #  244.069 M/sec
         490549944      branch-misses                    #    2.41% of all branches

      86.132655220 seconds time elapsed
      19.180209000 seconds user
      68.476075000 seconds sys
```

* Random check (ls -la filename) 30600 files into filesystem
Use perf stat ls -la <filename> to measure the query time for each file and sum up all elapsed times to calculate the total lookup cost.

Legacy :
min time: 0.00171 s
max time: 0.03799 s
avg time: 0.00423332 s
tot time: 129.539510 s

full_name_hash:
min time: 0.00171 s
max time: 0.03588 s
avg time: 0.00305601 s
tot time: 93.514040 s


<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Switched SimpleFS to hash-based directory indexing using a deterministic FNV-1a `simplefs_hash` to map filenames to extent/block slots, speeding up create/lookup/delete and readdir. On 30.6k files: create ~15% faster, delete ~19% faster, lookup ~28% faster; adds `mkfs.simplefs` support and CI on macOS.

- **New Features**
  - Hash-guided placement and lookup in create/link/symlink/rename/unlink with ring-scan wrap-around; fast `__file_lookup` shared by lookup/rename.
  - Faster traversal and cleanup: iterator skips fully scanned extents/blocks using `nr_files`; unlink/rename free empty directory extents; same-dir rename updates in place, cross-dir rename inserts then rolls back on failure.

- **Refactors and Fixes**
  - Added `hash.c` (FNV-1a with 64→32-bit fold) and utility macros (`CHECK_AND_SET_RING_INDEX`, `RELEASE_BUFFER_HEAD`); streamlined inode cache and allocation checks; clearer variable names; Makefile includes `hash.o`.
  - Fixed `simplefs_lookup` to scan all live entries (skip holes, bound by `SIMPLEFS_FILES_PER_BLOCK`); enforce filename max 254 chars (reserve NUL) and updated README.

<sup>Written for commit 22ee9e2c447d4b2d1b8ea2251cd4d5717911c7dd. Summary will update on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

jserv commented on symlink.c:
```diff
@@ -0,0 +1,14 @@
+
```
Remove this line.

RoyWFHuang replied:
removed

jserv commented on hash.c:
```diff
@@ -0,0 +1,12 @@
+        h *= 0x100000001b3ULL;
+    }
+    return h;
+}
```
Can you re-use the existing hash functions in Linux?

RoyWFHuang replied:
Add hash64 and full_name_hash as optional hash functions

visitorckw reviewed (changes requested):
I saw that your PR description includes some performance benchmarks, but the commit message lacks any performance numbers to support your improvements. Please improve the commit message.

visitorckw commented on bitmap.h:
```diff
@@ -48,11 +48,11 @@ static inline uint32_t get_free_blocks(struct super_block *sb, uint32_t len)
         return 0;
 
     sbi->nr_free_blocks -= len;
-    struct buffer_head *bh;
```
Please refrain from making extraneous or unrelated changes. Even if such changes are necessary, please split them into separate patches and explain in detail why they are needed in the commit message.

RoyWFHuang replied:
This change fixes the compiler error when building under the C90 standard, and I believe it doesn’t affect readability.
Separated this modification into a new commit.

visitorckw commented on super.c:
```diff
@@ -293,23 +293,21 @@ static journal_t *simplefs_get_dev_journal(struct super_block *sb,
     if (IS_ERR(bdev_file)) {
-        printk(KERN_ERR
-               "failed to open journal device unknown-block(%u,%u) %ld\n",
+        pr_err("failed to open journal device unknown-block(%u,%u) %ld\n",
```
Why do we need to change this to use pr_err()?
This also appears to be an unrelated change.

RoyWFHuang replied:
This change is to align the Simplefs print function.

RoyWFHuang replied:
Created a new commit for these adjustments.

visitorckw commented on dir.c:
```diff
@@ -62,9 +62,10 @@ static int simplefs_iterate(struct file *dir, struct dir_context *ctx)
         /* Iterate over blocks in one extent */
-        for (bi = 0; bi < eblock->extents[ei].ee_len && remained_nr_files;
+        for (bi = 0;
+             bi < eblock->extents[ei].ee_len && remained_nr_files && ei_nr;
```
IIUC, this optimizes efficiency by using nr_files to break out of the loop early, but it is unrelated to "hash lookup". It should be a separate patch. Please follow the "One Logical Change Per Patch" principle and split your changes.

RoyWFHuang replied:
Yes, this change is to boost nr_blk utilization. It is a legacy issue by a previous commit, and I separated it from the hash commit.

RoyWFHuang committed:
Optimize the file iteration

Each `simplefs_extent` structure contains a counter that records the
total number of files within that extent. When the counter matches the
expected file number, it indicates there are no more files after this
index, allowing the iterator to skip directly to the next extension block.
This reduces unnecessary scanning and improves traversal efficiency.

jserv commented on .config:
```diff
@@ -0,0 +1,11 @@
+#
+# CONFIG_SIMPLEFS_HASH_HASH64 is not set
+CONFIG_SIMPLEFS_HASH_FNV1A=y
+# CONFIG_SIMPLEFS_HASH_FULLNAME is not set
```
Don't check in this.

RoyWFHuang replied:
Remove it and add default into makefile.

RoyWFHuang commented:
> How can you determine which hash function is the most suitable?

I’m not sure if "fnv" is the most suitable, but index in SimpleFS is relatively small, using a more complex algorithm might not provide significant benefits. I think fnv is a reasonable balance between simplicity and performance.

RoyWFHuang commented on simplefs.h:
```diff
@@ -119,6 +119,7 @@ struct dentry *simplefs_mount(struct file_system_type *fs_type,
 extern const struct file_operations simplefs_file_ops;
 extern const struct file_operations simplefs_dir_ops;
 extern const struct address_space_operations simplefs_aops;
+static const struct vm_operations_struct simplefs_file_vm_ops;
```
remove it.

visitorckw reviewed (changes requested):
You ignored many of my comments without making any changes or providing any replies. You still retained many irrelevant changes, making the review difficult. Additionally, a single-line commit message saying only "optimize the file search process" is way too vague. Please improve the git commit message.

visitorckw commented on hash.c:
```diff
@@ -0,0 +1,40 @@
+
+
+#if defined(CONFIG_SIMPLEFS_HASH_HASH64)
+static uint64_t _hash64(const char *name, size_t len)
```
Is it really necessary for us to re-implement our own hash function? Upstream Linux already has many hash implementations, like full_name_hash(). Could we use those here instead?

RoyWFHuang replied:
As I mentioned earlier, the existing hash functions in the Linux kernel may not be suitable for Simplefs because its index space is relatively small. We need a simple and fast hash function to compute the hash index.

RoyWFHuang committed:
Rename variables for better readability

jserv commented on Makefile:
```diff
@@ -22,6 +26,30 @@ $(IMAGE): $(MKFS)
+
+.config:
+	@echo "Creating default .config"
+	@echo "CONFIG_SIMPLEFS_HASH_FNV1A=y" > .config
```
Can you avoid changing the configurations?

jserv commented on hash.c:
```diff
@@ -0,0 +1,40 @@
+    }
+    return h;
+}
+#endif
```
It is not worthwhile to make the hash function configurable at build time. Instead, simply use the one provided by the Linux kernel or whichever implementation proves most effective.

RoyWFHuang replied:
Remove other hash options and retain only `full_name_hash` for indexing.

jserv commented on .github/workflows/main.yaml:
```diff
@@ -9,7 +9,7 @@ jobs:
     - name: install-dependencies
       run: |
-            sudo apt-get install build-essential
+            sudo apt-get install build-essential cmake-kconfig
```
Don't install `cmake-kconfig`, which is not really needed since only one hash function is required.

RoyWFHuang replied:
Removed.

visitorckw reviewed (changes requested):
Quoted from patch 2:
> Align the print function with the Simplefs print format for consistency.
> Also adjust variable declarations to fix compiler warnings when building
> under the C90 standard.

I'm unsure which Linux kernel versions simplefs currently intends to support, but AFAIK, the Linux kernel currently uses gnu c11 as its standard.

Furthermore, the word "Also" is often a sign that the change should be in a separate patch. In my view, you are performing two distinct actions here:

a) Changing printk -> pr_err.
b) Fixing a compiler warning.

I also remain confused as to whether the printk to pr_err change is truly warranted, and what relevance it has to the PR's title, which is "Use hash func to boost file creation and lookup".

RoyWFHuang commented:
> Quoted from patch 2:
> 
> > Align the print function with the Simplefs print format for consistency.
> > Also adjust variable declarations to fix compiler warnings when building
> > under the C90 standard.
> 
> I'm unsure which Linux kernel versions simplefs currently intends to support, but AFAIK, the Linux kernel currently uses gnu c11 as its standard.
> 
SimpleFS targets kernel 5.10, and since 5.10 recommends using C11, this commit is no longer needed. I have removed it.

> Furthermore, the word "Also" is often a sign that the change should be in a separate patch. In my view, you are performing two distinct actions here:
> 
> a) Changing printk -> pr_err. b) Fixing a compiler warning.
> 
> I also remain confused as to whether the printk to pr_err change is truly warranted, and what relevance it has to the PR's title, which is "Use hash func to boost file creation and lookup".

OK, I agree. I’ve removed these commits. If needed, I will create another PR to address it.

As far as I know, the pr_info rule was already established in the initial commit, and printk appeared only in later changes. I think we should follow the original rule and keep it consistent.

RoyWFHuang commented on inode.c:
```diff
@@ -741,186 +816,179 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
+    goto release_new;
+
+rm_new:
+    simplefs_remove_from_dir(dest_dir, dest_dentry, &dest_ei);
```
This “simplefs_remove_from_dir” only removes the metadata from the destination directory. It does not clean dentry data.

RoyWFHuang commented on inode.c:
```diff
@@ -741,186 +816,179 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
+    if (!eblk_src->extents[src_ei].nr_files) {
+        put_blocks(sbi, eblk_src->extents[src_ei].ee_start,
+                   eblk_src->extents[src_ei].ee_len);
+        memset(&eblk_src->extents[src_ei], 0, sizeof(struct simplefs_extent));
```
Fixed it.

RoyWFHuang commented on inode.c:
```diff
@@ -110,6 +110,74 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+    /* Read the directory block on disk */
+    bh = sb_bread(sb, ci_dir->ei_block);
+    if (!bh)
+        return false;
```
Add a return error code and modify the related functions.

RoyWFHuang commented on inode.c:
```diff
@@ -741,186 +816,179 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
+        if (dblock->nr_files != SIMPLEFS_FILES_PER_BLOCK) {
+            simplefs_set_file_into_dir(dblock, src_in->i_ino,
+                                       dest_dentry->d_name.name);
+            eblk_dest->extents[dest_ei].nr_files++;
```
Fixed

visitorckw commented on hash.c:
```diff
@@ -0,0 +1,9 @@
+{
+    return full_name_hash(dentry, dentry->d_name.name,
+                          strlen(dentry->d_name.name));
+}
```
So this C file contains only a single function, which effectively just calls full_name_hash() directly. Do we really need this extra API? Also, why is the name simple_hash() considered better than full_name_hash()?

RoyWFHuang replied:
This follows the file structure used by the ext4 filesystem. If we need to switch to another algorithm in the future, we can simply replace the implementation inside the simple_hash() function. Using this interface allows us to separate the actual implementation from the rest of the code.

If you think this abstraction is unnecessary, we can inline it instead.

visitorckw commented on dir.c:
```diff
@@ -62,9 +62,10 @@ static int simplefs_iterate(struct file *dir, struct dir_context *ctx)
         /* Iterate over blocks in one extent */
-        for (bi = 0; bi < eblock->extents[ei].ee_len && remained_nr_files;
+        for (bi = 0;
+             bi < eblock->extents[ei].ee_len && remained_nr_files && ei_nr;
```
How often does ei_nr actually hit 0 during the loop check? Is the overhead of maintaining ei_nr really justified by the benefit of terminating the loop early?

RoyWFHuang replied:
full_name_hash in remove file
```
          83278.61 msec task-clock                       #    0.967 CPUs utilized
             52431      context-switches                 #  629.585 /sec
              1309      cpu-migrations                   #   15.718 /sec
           3796501      page-faults                      #   45.588 K/sec
      199894058328      cycles                           #    2.400 GHz
      110625460371      instructions                     #    0.55  insn per cycle
       20325767251      branches                         #  244.069 M/sec
         490549944      branch-misses                    #    2.41% of all branches

      86.132655220 seconds time elapsed
      19.180209000 seconds user
      68.476075000 seconds sys

remove ei_nr:
         136413.23 msec task-clock                       #    0.962 CPUs utilized
             54993      context-switches                 #  403.135 /sec
              2504      cpu-migrations                   #   18.356 /sec
           3827233      page-faults                      #   28.056 K/sec
      323663184097      cycles                           #    2.373 GHz
      154283317313      instructions                     #    0.48  insn per cycle
       30531054743      branches                         #  223.813 M/sec
         561254005      branch-misses                    #    1.84% of all branches

     141.745463391 seconds time elapsed
      21.300616000 seconds user
     120.011093000 seconds sys

```

full_name_hash in list single file
min time: 0.00171 s
max time: 0.03588 s
avg time: 0.00305601 s
tot time: 93.514040 s

remove ei_nr:
total test 30600
min: 0.0017 s
max: 0.0432 s
avg: 0.00342463 s
tot: 104.793540 s

visitorckw commented:
Hmmm, I see my comments were marked as resolved, but I don't see any changes or replies. Since you re-requested a review, did you perhaps forget to push the new changes, or am I missing something?

jserv commented on inode.c:
```diff
@@ -126,51 +197,32 @@ static struct dentry *simplefs_lookup(struct inode *dir,
+        if (!chk)
+            return ERR_PTR(-chk);
+        goto search_end;
+    }
```
error handling is broken.

`ERR_PTR(-0)` returns NULL, which is a valid negative dentry lookup result. This silently converts "not found" to success with NULL inode. The logic is inverted and confusing.

Proposed fix:
```c
  chk = __file_lookup(dir, dentry, &ei, &bi, &fi);
  if (chk < 0)
      return ERR_PTR(chk);  // I/O error
  if (chk == 0)
      goto search_end;  // Not found, return NULL dentry
```

RoyWFHuang replied:
Fixed it and also used kernel error code for checking

jserv commented on inode.c:
```diff
@@ -110,6 +110,77 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+    /* Check filename length */
+
+    /* Read the directory block on disk */
+    bh = sb_bread(sb, ci_dir->ei_block);
```
When sb_bread fails in __file_lookup, ret = -EIO is set but the function returns ret (negative), which the caller then misinterprets. The error path in simplefs_lookup doesn't properly propagate I/O errors.

RoyWFHuang replied:
Fix the return value and checking the error code.

jserv commented on inode.c:
```diff
@@ -110,6 +110,77 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+
+    for (idx_ei = 0; idx_ei < SIMPLEFS_MAX_EXTENTS; _ei++, idx_ei++) {
+        if (unlikely(_ei >= SIMPLEFS_MAX_EXTENTS))
+            _ei %= SIMPLEFS_MAX_EXTENTS;
```
The wraparound search (`_ei %= SIMPLEFS_MAX_EXTENTS`) handles collisions, but the worst case is still O(n) if many files hash to the same bucket.

Consider documenting expected collision rates with your benchmark files.

jserv replied:
This pattern repeats 6+ times. A helper function would reduce the 300+ line.

jserv commented on inode.c:
```diff
@@ -110,6 +110,77 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+                         struct dentry *dentry,
+                         int *ei,
+                         int *bi,
+                         int *fi)
```
Using 0/1/negative is confusing. Standard kernel patterns:
- Return 0 on success, negative errno on failure
- Use output parameter for "found/not found" boolean

RoyWFHuang replied:
Use kernel-defined error codes for return values.

jserv reviewed (changes requested):
Rebase the latest `master` branch.

jserv commented on inode.c:
```diff
@@ -110,69 +110,118 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+    eblock = (struct simplefs_file_ei_block *) bh->b_data;
+    bh2 = sb_bread(sb, eblock->extents[ei].ee_start + bi);
+    if (!bh2)
+        return ERR_PTR(-EIO);
```
Buffer head leak in `simplefs_lookup()`: When `bh2` read fails, `bh` is never released.

jserv commented on inode.c:
```diff
@@ -753,186 +833,187 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
+    goto release_new;
+
+rm_new:
+    simplefs_remove_from_dir(dest_dir, dest_dentry, &dest_ei);
```
`dest_dentry` was never inserted - it has no inode. `simplefs_remove_from_dir()` calls `d_inode(dentry)` unconditionally. Kernel panic on any rename
  error after `simplefs_put_new_ext()`.

jserv commented on inode.c:
```diff
@@ -110,69 +110,118 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
-        return ERR_PTR(-EIO);
+        return -EIO;
     eblock = (struct simplefs_file_ei_block *) bh->b_data;
+    hash_code = simplefs_hash(dentry) %
+                (SIMPLEFS_MAX_EXTENTS * SIMPLEFS_MAX_BLOCKS_PER_EXTENT);
```
`full_name_hash()` is designed for VFS dentry caching, not for on-disk indexing. It uses a per-boot salt that changes on reboot. Consider: what happens to hash placement after reboot?

If `full_name_hash()` returns different values after reboot (it does), lookups will start from wrong positions. The linear probe fallback saves correctness, but destroys the performance benefit.

jserv replied:
Use a deterministic hash like FNV-1a or djb2.

jserv commented:
@cubic-dev-ai Continue analyzing and warning until:
1. Fix all buffer head leaks
2. Add missing mark_buffer_dirty() calls
3. Fix the rename error path panic
4. Extract wraparound logic into helper function

jserv commented on inode.c:
```diff
@@ -11,6 +11,11 @@
 static const struct inode_operations simplefs_inode_ops;
 static const struct inode_operations symlink_inode_ops;
 
+#define CHECHK_AND_SET_RING_INDEX(idx, len) \
```
should be `CHECK_AND_SET_RING_INDEX` (typo)

RoyWFHuang replied:
Fixed

jserv commented on inode.c:
```diff
@@ -11,6 +11,11 @@
+#define CHECHK_AND_SET_RING_INDEX(idx, len) \
+    do {                                    \
+        if (unlikely(idx >= len))           \
+            idx %= len;                     \
```
Modulo operation on every wraparound - could use `if (idx >= len) idx -= len` for single wraparound case.

RoyWFHuang replied:
Done

jserv commented on inode.c:
```diff
@@ -753,186 +848,223 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
+
+    goto release_new;
+
+rm_new:
+    if (dest_inserted) {
+        bh_ext = sb_bread(
+            sb, eblk_dest->extents[dest_ei].ee_start + dest_inserted_bi);
+        if (bh_ext) {
+            dblock = (struct simplefs_dir_block *) bh_ext->b_data;
+            if (simplefs_try_remove_entry(dblock, eblk_dest, dest_ei,
+                                          src_in->i_ino,
+                                          dest_dentry->d_name.name)) {
+                mark_buffer_dirty(bh_ext);
+                mark_buffer_dirty(bh_fei_blk_dest);
+            }
+            brelse(bh_ext);
+        }
```
If `simplefs_try_remove_entry` fails (I/O error), do we leak the destination entry?
The code doesn't check return value.

RoyWFHuang replied:
Use the dest_bh_ext cache in the insert section to quickly find deletion targets.

jserv commented on hash.c:
```diff
@@ -0,0 +1,16 @@
+uint32_t simplefs_hash(struct dentry *dentry)
+{
+    /* FIX: Use a deterministic hash like FNV-1a or djb2.*/
+    /* Use fnv1a_64 algorithm */
```
Confusing comment. The code DOES use FNV-1a. Remove "FIX:" or clarify intent.

RoyWFHuang replied:
Removed

jserv commented on inode.c:
```diff
@@ -563,61 +653,55 @@ static int simplefs_create(struct inode *dir,
 
     eblock = (struct simplefs_file_ei_block *) bh->b_data;
+    dir_nr_files = eblock->nr_files;
+
```
Drop an unintended blank line.

jserv commented on inode.c:
```diff
@@ -753,186 +848,223 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
+        dblock = (struct simplefs_dir_block *) bh_ext->b_data;
+
+        strncpy(dblock->files[fi].filename, dest_dentry->d_name.name,
+                SIMPLEFS_FILENAME_LEN);
```
File becomes unfindable after rename because:
- Old entry stays at hash(old_name) location
- Future lookups calculate hash(new_name) and search wrong bucket
- File effectively vanishes from namespace

RoyWFHuang committed:
Use hash func to boost file creation and lookup

Introduce a hash-based mechanism to speed up file creation and lookup
operations. The hash function enables faster access to extent and logical
block extent index, improving overall filesystem performance.

hash_code = file_hash(file_name);
extent index = hash_code / SIMPLEFS_MAX_BLOCKS_PER_EXTENT
block index = hash_code % SIMPLEFS_MAX_BLOCKS_PER_EXTENT;

Use perf to measure:
1. File Creation (random)
Legacy:
259.842753513 seconds time elapsed
23.000247000 seconds user
150.380145000 seconds sys

full_name_hash:
222.274028604 seconds time elapsed
20.794966000 seconds user
151.941876000 seconds sys

2. File Listing (random)
Legacy:
min time: 0.00171 s
max time: 0.03799 s
avg time: 0.00423332 s
tot time: 129.539510 s

full_name_hash:
min time: 0.00171 s
max time: 0.03588 s
avg time: 0.00305601 s
tot time: 93.514040 s

3. files Removal (Random)
Legacy:
106.921706288 seconds time elapsed
16.987883000 seconds user
91.268661000 seconds sys

full_name_hash:
86.132655220 seconds time elapsed
19.180209000 seconds user
68.476075000 seconds sys

jserv commented on inode.c:
```diff
@@ -11,6 +11,11 @@
 static const struct inode_operations simplefs_inode_ops;
 static const struct inode_operations symlink_inode_ops;
 
+#define CHECK_AND_SET_RING_INDEX(idx, len) \
```
`CHECK_AND_SET_RING_INDEX` is not a proper ring wrap. Single subtraction, not modulo. If `idx` starts at 7 (from `hash % SIMPLEFS_MAX_BLOCKS_PER_EXTENT`) but the current extent has `ee_len = 2`, the macro yields `7 - 2 = 5` -- still out of bounds. `sb_bread` then reads a block outside the extent.

Use `idx %= len`, or at minimum clamp the initial value to the actual extent length before entering the inner loop.

RoyWFHuang replied:
In the current simplefs implementation, the hash_code is restricted to be within SIMPLEFS_MAX_EXTENTS * SIMPLEFS_MAX_BLOCKS_PER_EXTENT. Following this logic:

ei = hash_code / SIMPLEFS_MAX_BLOCKS_PER_EXTENT ensures ei stays within 0 to SIMPLEFS_MAX_EXTENTS - 1.

bi = hash_code % SIMPLEFS_MAX_BLOCKS_PER_EXTENT ensures bi stays within 0 to SIMPLEFS_MAX_BLOCKS_PER_EXTENT - 1.

Therefore, if we apply the mapping correctly using:
CHECK_AND_SET_RING_INDEX(ei, SIMPLEFS_MAX_EXTENTS);
CHECK_AND_SET_RING_INDEX(_bi, SIMPLEFS_MAX_BLOCKS_PER_EXTENT);

We should not encounter out-of-bounds errors under normal conditions."

jserv commented on inode.c:
```diff
@@ -110,6 +115,75 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+            continue;
+
+        /* Iterate blocks in extent */
+        for (idx_bi = 0; idx_bi < eblock->extents[_ei].ee_len;
```
`_bi` is never reset when moving to a new extent. It is initialized once from the hash. When the outer `_ei` loop advances to the next extent, `_bi` carries over from the previous extent's iteration. The inner loop scans wrong blocks or skips the block that actually contains the target file.

Reset `_bi = 0` (or re-derive it) at each new extent boundary. Same bug exists in `simplefs_remove_from_dir` at line ~688.

RoyWFHuang replied:
Fixed.

jserv commented on inode.c:
```diff
@@ -110,6 +115,75 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+
+            dblock = (struct simplefs_dir_block *) bh2->b_data;
+            /* Search file in ei_block */
+            for (_fi = 0; _fi < dblock->nr_files;) {
```
Unchecked `nr_blk` from disk -- if on-disk `nr_blk` is 0, `_fi += dblock->files[_fi].nr_blk` loops forever. If `nr_blk` is oversized, `_fi` jumps past `SIMPLEFS_FILES_PER_BLOCK` and indexes out of bounds on the next iteration.

Validate `nr_blk >= 1 && _fi + nr_blk <= SIMPLEFS_FILES_PER_BLOCK` before advancing; return `-EIO` on corrupt metadata.

RoyWFHuang replied:
I believe this situation should not occur under normal conditions, as nr_blk (or nr_files) should always be at least 1. If we encounter a 0, it likely indicates a bug in the block merging or splitting logic.

jserv commented on inode.c:
```diff
@@ -126,52 +200,38 @@ static struct dentry *simplefs_lookup(struct inode *dir,
+    if (chk < 0)
+        return ERR_PTR(chk); /* I/O error */
+
     bh = sb_bread(sb, ci_dir->ei_block);
```
Redundant I/O -- `__file_lookup` already reads the ei_block and dir_block to find the file, then releases them. Here `simplefs_lookup` reads them *again* just to extract the inode number. This doubles the I/O on every successful lookup and partially defeats the hash optimization.

Have `__file_lookup` return the inode number directly instead of the (ei, bi, fi) triple.

RoyWFHuang replied:
Regarding the redundant I/O:

The current implementation of `__file_lookup` is designed only to locate the (ei, bi, fi) indices and does not maintain the bh cache. Returning the inode number directly or keeping the cache active would require refactoring all related functions and their call sites. 

I agree this is a valid performance concern. I plan to create a dedicated issue to track this enhancement and implement a more efficient caching mechanism in a future update.

Additionally, regarding file search performance, we can leverage the "nr_files" information to further boost search speed. I will include this optimization in the same roadmap to improve overall directory operation efficiency.

jserv commented on inode.c:
```diff
@@ -401,6 +460,39 @@ static void simplefs_set_file_into_dir(struct simplefs_dir_block *dblock,
+                !strcmp(dblock->files[fi].filename, name)) {
+                dblock->files[fi].inode = 0;
+                /* Merge the empty data */
+                for (i = fi - 1; i >= 0; i--) {
```
Backward merge fails when `fi == 0`. The loop `for (i = fi - 1; i >= 0; i--)` starts at `i = -1` and immediately exits. The freed space at slot 0 is never coalesced with adjacent free space. Over time this leaks directory entry slots.

Handle the `fi == 0` case explicitly -- the freed slot becomes the head of the free chain.

RoyWFHuang replied:
I believe this case is already handled. We should separate this issue into two scenarios based on the state of the directory:

1. If fi = 0 is a file and fi = 1 is empty:
In this case, the nr_blk for fi = 0 should be >= 2 (as nr_blk is used to record contiguous empty blocks). When the file at fi = 0 is removed, we only need to update the metadata to reflect that the slot is now part of the free space.

2. If fi = 0 is a file and fi = 1 also contains a file:
The nr_blk would be 1. When the file at fi = 0 is removed, it starts recording contiguous empty blocks from that position. Since the following slot (fi = 1) is occupied, the logic remains consistent and doesn't require an explicit change for the fi = 0 case.

To clarify, nr_blk has two different meanings in our implementation:
- If a block is empty: It indicates the number of contiguous empty blocks.
- If the leading block contains a file: It indicates the total span, including the used block and any trailing empty space.

jserv commented on inode.c:
```diff
@@ -433,12 +525,12 @@ static int simplefs_create(struct inode *dir,
 #endif
-    int ret = 0, alloc = false;
-    int bi = 0;
+    int ret = 0, alloc = false, hash_code;
```
`hash_code` declared `int` but `simplefs_hash()` returns `uint32_t`. Mixing signed/unsigned weakens the type contract and invites bugs on future refactors. Declare as `uint32_t`.

RoyWFHuang replied:
Fixed

jserv commented on inode.c:
```diff
@@ -753,186 +852,233 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
         }
+        dblock = (struct simplefs_dir_block *) bh_ext->b_data;
+
+        strncpy(dblock->files[fi].filename, dest_dentry->d_name.name,
```
In-place rename in full directory breaks the hash invariant. When `src_dir == dest_dir` and the directory is full, this just overwrites the filename in the old slot. But the entry stays at the hash position of the *old* name. Future lookups hash the *new* name to a different bucket and will not find the file.

This makes renamed files unreachable by normal lookup/delete. Remove the in-place shortcut; do remove-then-reinsert using the new name's hash, or return an error if the target bucket is full.

RoyWFHuang replied:
I think an in-place rename is sufficient in this case. If src_dir == dest_dir and the directory is full, even if we were to remove the old entry and re-insert it, the new entry would likely end up in the same slot anyway. 

Even if the new filename's hash doesn't match the current slot, the file remains reachable. Our lookup mechanism performs a linear search starting from the hash index and continues through the ring until it finds the entry or returns to the start. Therefore, modifying the filename in-place avoids the overhead of a full remove-and-reinsert operation without sacrificing correctness.

jserv commented on inode.c:
```diff
@@ -1021,8 +1167,10 @@ static int simplefs_link(struct dentry *old_dentry,
+
+    hash_code = simplefs_hash(dentry) %
+                (SIMPLEFS_MAX_EXTENTS * SIMPLEFS_MAX_BLOCKS_PER_EXTENT);
+    avail = simplefs_get_available_ext_idx(eblock, hash_code);
```
Error path frees wrong extent. The `put_block:` label at line 1228 frees `eblock->extents[ei]`, but `ei` is still 0 (initialized at declaration, never updated). The actually allocated extent is at index `avail`. On I/O failure after `alloc = true`, extent 0 gets freed/zeroed instead of the new one.

Use `avail` in the `put_block:` cleanup path, matching what `simplefs_create` already does. Same bug in `simplefs_symlink` at line 1349.

RoyWFHuang replied:
Fixed.

jserv commented on hash.c:
```diff
@@ -0,0 +1,15 @@
+
+uint32_t simplefs_hash(struct dentry *dentry)
+{
+    const char *str = dentry->d_name.name;
```
Unkeyed FNV-1a is vulnerable to HashDoS. FNV-1a is deterministic and trivially reversible. Attackers can craft filenames that all collide, forcing O(n) linear scans on every directory operation.

Consider using the kernel's `full_name_hash()` which is SipHash-based with a per-boot random key, or at minimum incorporate a per-superblock random salt.

RoyWFHuang replied:
Based on our previous discussion

```
full_name_hash() is designed for VFS dentry caching, not for on-disk indexing. It uses a per-boot salt that changes on reboot. Consider: what happens to hash placement after reboot?

If full_name_hash() returns different values after reboot (it does), lookups will start from wrong positions. The linear probe fallback saves correctness, but destroys the performance benefit.

Use a deterministic hash like FNV-1a or djb2.
```

I believe we should stick with this approach to ensure consistency across reboots.

jserv commented on inode.c:
```diff
@@ -563,61 +659,54 @@ static int simplefs_create(struct inode *dir,
+    bi = hash_code % SIMPLEFS_MAX_BLOCKS_PER_EXTENT;
+
+    for (; dir_nr_files; ei++) {
+        CHECK_AND_SET_RING_INDEX(ei, SIMPLEFS_MAX_EXTENTS);
```
Same `_bi` carry-across-extent bug as `__file_lookup`. `bi` is seeded from the hash once at line ~687 and then carried across extent iterations without reset. When `ei` advances, the inner loop starts scanning from a stale `bi` value that belongs to the previous extent. Files can become undeletable.

Reset `bi` when entering a new extent.

RoyWFHuang replied:
Fixed.

jserv reviewed (changes requested):
The hash-based directory indexing idea is sound and the perf numbers are real. But the implementation has several correctness bugs that can cause data corruption or make files unreachable. Three blocking issues found independently by all reviewers.

Minor: orphan comment (line 133), double blank lines (161, 1170), stale TODOs (1195, 1312), `printk` should be `pr_info` (1166, 1283), typo "founded" -> "found" (911), commit message typo "fiile".

Needs a v2 to fix the data-corruption and unreachable-file bugs before merge.

jserv commented on inode.c:
```diff
@@ -656,9 +749,21 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
         return ret;
+    bh = sb_bread(sb, SIMPLEFS_INODE(dir)->ei_block);
+    if (!bh)
+        return -EIO;
```
**CRITICAL — orphans the inode on I/O failure**

```c
ret = simplefs_remove_from_dir(dir, dentry, &ei);
if (ret != 0)
    return ret;
bh = sb_bread(sb, SIMPLEFS_INODE(dir)->ei_block);
if (!bh)
    return -EIO;          /* dirent already gone, inode + blocks leak */
```

The directory entry has already been removed and dirtied by `simplefs_remove_from_dir`. Returning `-EIO` here leaves the inode allocated with no name pointing at it — userspace sees `EIO` and the inode/blocks are permanently leaked.

**Fix:** move the empty-extent reclamation inside `simplefs_remove_from_dir`, where the same `bh` is still held. No second `sb_bread` needed.

RoyWFHuang replied:
Fixed

jserv commented on inode.c:
```diff
@@ -753,186 +856,233 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
+    if (!bh_fei_blk_src) {
+        ret = -EIO;
+        goto rm_new;
+    }
```
**CRITICAL — rename rollback orphans the file**

```c
ret = simplefs_remove_from_dir(src_dir, src_dentry, &src_ei);
if (ret != 0)
    goto rm_new;          /* rolls back dest insert -- ok */

bh_fei_blk_src = sb_bread(sb, ci_src->ei_block);
if (!bh_fei_blk_src) {
    ret = -EIO;
    goto rm_new;          /* WRONG: src already removed */
}
```

After `simplefs_remove_from_dir(src)` succeeds, the source dirent is gone from disk. If the next `sb_bread` fails, `rm_new` removes the dest entry too. The file now has no name in either directory — orphaned, blocks leaked.

**Fix:** read any source metadata you need *before* removing the src entry, or do empty-extent reclamation inside `simplefs_remove_from_dir` like the unlink fix above.

RoyWFHuang replied:
Fixed

jserv commented on inode.c:
```diff
@@ -110,6 +115,77 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+    _ei = hash_code / SIMPLEFS_MAX_BLOCKS_PER_EXTENT;
+    _bi = hash_code % SIMPLEFS_MAX_BLOCKS_PER_EXTENT;
+    int nr_ei_files = eblock->nr_files;
+    for (idx_ei = 0; nr_ei_files; _ei++, idx_ei++) {
```
**CRITICAL — unbounded ring loop on corrupt metadata**

```c
for (idx_ei = 0; nr_ei_files; _ei++, idx_ei++) {
    CHECK_AND_SET_RING_INDEX(_ei, SIMPLEFS_MAX_EXTENTS);
    if (!eblock->extents[_ei].ee_start)
        continue;          /* nr_ei_files unchanged: spin forever */
    nr_ei_files -= eblock->extents[_ei].nr_files;
    ...
}
```

On a directory whose `eblock->nr_files != 0` but all `extents[].ee_start == 0` (corruption), this never terminates — kernel hang in process context. Never trust on-disk metadata as a sole loop bound.

**Fix:** add `idx_ei < SIMPLEFS_MAX_EXTENTS` to the loop condition. Validate `extents[_ei].nr_files <= nr_ei_files` before subtracting, return `-EFSCORRUPTED` on inconsistency.

RoyWFHuang replied:
In our design, `eblock->nr_files` records the total number of files within a `simplefs_file_ei_block`. This value is expected to be identical to the sum of `nr_files` across all extent blocks (`eblock->extents[_ei].nr_files`). If these values do not match, it would indicate a bug in our file creation or management logic.

However, I acknowledge the importance of defensive programming when dealing with on-disk metadata to prevent potential kernel hangs. I will create a dedicated issue to improve these validation checks and ensure the file system's robustness against metadata corruption.

jserv commented on inode.c:
```diff
@@ -110,6 +115,77 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+        nr_ei_files -= eblock->extents[_ei].nr_files;
+        /* Iterate blocks in extent */
+        int nr_bi_files = eblock->extents[_ei].nr_files;
+        for (idx_bi = 0; nr_bi_files; _bi++, idx_bi++) {
```
**CRITICAL — same unbounded-loop pattern at the inner block level**

```c
int nr_bi_files = eblock->extents[_ei].nr_files;
for (idx_bi = 0; nr_bi_files; _bi++, idx_bi++) {
    ...
    nr_bi_files -= dblock->nr_files;   /* signed int, can go negative */
}
```

A corrupt `dblock->nr_files == 0` keeps `nr_bi_files` constant; an oversized value drives it negative, which is still truthy. Kernel hang.

**Fix:** `idx_bi < eblock->extents[_ei].ee_len`; reject `dblock->nr_files > SIMPLEFS_FILES_PER_BLOCK` and return `-EFSCORRUPTED`.

RoyWFHuang replied:
Same as above

jserv commented on inode.c:
```diff
@@ -563,65 +662,59 @@ static int simplefs_create(struct inode *dir,
+    ei = hash_code / SIMPLEFS_MAX_BLOCKS_PER_EXTENT;
+    bi = hash_code % SIMPLEFS_MAX_BLOCKS_PER_EXTENT;
+
+    for (; dir_nr_files; ei++) {
```
**CRITICAL — `simplefs_remove_from_dir` repeats the unbounded-ring pattern**

```c
for (; dir_nr_files; ei++) {
    CHECK_AND_SET_RING_INDEX(ei, SIMPLEFS_MAX_EXTENTS);
    if (eblock->extents[ei].ee_start) { ... }
    bi = 0;
}
```

When `extents[ei].ee_start == 0`, the body is skipped and `dir_nr_files` is never decremented. Same hang risk as `__file_lookup`.

**Fix:** bound the outer scan with a separate counter `< SIMPLEFS_MAX_EXTENTS` and validate counters before subtracting.

RoyWFHuang replied:
Same as above

jserv commented on inode.c:
```diff
@@ -110,6 +115,77 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+                    brelse(bh2);
+                    return 0;
+                }
+                _fi += dblock->files[_fi].nr_blk;
```
**HIGH — unchecked `nr_blk` in run-length walk**

```c
_fi += dblock->files[_fi].nr_blk;
```

On-disk `nr_blk == 0` → infinite loop. Oversized `nr_blk` → next iteration indexes past the block. Same code pattern exists in `simplefs_try_remove_entry` line 493.

**Fix:** validate `nr_blk >= 1 && _fi + nr_blk <= SIMPLEFS_FILES_PER_BLOCK` before stepping; return `-EFSCORRUPTED` on violation.

RoyWFHuang replied:
Regarding the concerns about `nr_blk == 0`, potential infinite loops, and metadata validation (including line 493 in `simplefs_try_remove_entry`):

As I mentioned previously:
"I believe this situation should not occur under normal conditions, as `nr_blk` (or `nr_files`) should always be at least 1. If we encounter a 0, it likely indicates a bug in the block merging or splitting logic."

While I understand the concern regarding out-of-bounds indexing or infinite loops in the case of disk corruption, I plan to implement these defensive programming measures and error handling as part of a separate, dedicated metadata optimization task. 

This current PR is strictly focused on the core hash-based indexing improvements and the resulting performance gains. To maintain a clean and focused commit history, I believe it is better to handle these robustness enhancements in a subsequent PR, ensuring the code review process remains efficient for both distinct functional changes.

jserv commented on inode.c:
```diff
@@ -401,6 +462,39 @@ static void simplefs_set_file_into_dir(struct simplefs_dir_block *dblock,
+            }
+            blk_nr_files--;
+        }
+        fi += dblock->files[fi].nr_blk;
```
**HIGH — same unchecked `nr_blk` walk as `__file_lookup`**

```c
fi += dblock->files[fi].nr_blk;
```

Mirror of the bug at line 178. Validate `nr_blk >= 1 && fi + nr_blk <= SIMPLEFS_FILES_PER_BLOCK`.

RoyWFHuang replied:
Same as above
https://github.com/sysprog21/simplefs/pull/79#discussion_r3045719059

jserv reviewed (commented):
The hash-based directory indexing direction is right and the benchmark numbers are real, but the orphan paths and unbounded ring loops are merge blockers. Findings posted as inline comments below.

jserv commented on inode.c:
```diff
@@ -753,186 +856,233 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
+        dblock = (struct simplefs_dir_block *) bh_ext->b_data;
+
+        strncpy(dblock->files[fi].filename, dest_dentry->d_name.name,
+                SIMPLEFS_FILENAME_LEN);
```
**HIGH — in-place rename loses filename NUL terminator**

```c
strncpy(dblock->files[fi].filename, dest_dentry->d_name.name,
        SIMPLEFS_FILENAME_LEN);
```

Every other place that writes a filename uses `SIMPLEFS_FILENAME_LEN - 1` followed by an explicit `'\0'` (e.g. `simplefs_set_file_into_dir` at lines 449-451, 455-456, 459-460). This call site doesn't. A 255-byte filename leaves the buffer un-terminated, corrupting subsequent `strcmp`/`strncmp` results in lookup and remove paths.

**Fix:** match the existing pattern:
```c
strncpy(dblock->files[fi].filename, dest_dentry->d_name.name,
        SIMPLEFS_FILENAME_LEN - 1);
dblock->files[fi].filename[SIMPLEFS_FILENAME_LEN - 1] = '\0';
```

RoyWFHuang replied:
Regarding the `strncpy` filename copy:

I recall that in the original design, the maximum filename length was intended to be 255. If we modify this to `SIMPLEFS_FILENAME_LEN - 1` to ensure a null-terminator, I believe we also need to update the filename length validation in the file creation path accordingly to maintain consistency. 

I will synchronize these changes.

jserv commented on inode.c:
```diff
@@ -126,52 +202,38 @@ static struct dentry *simplefs_lookup(struct inode *dir,
+    if (chk < 0)
+        return ERR_PTR(chk); /* I/O error */
+
     bh = sb_bread(sb, ci_dir->ei_block);
```
**HIGH — redundant double-read defeats the PR's perf goal**

```c
chk = __file_lookup(dir, dentry, &ei, &bi, &fi);
...
bh  = sb_bread(sb, ci_dir->ei_block);              /* re-read ei_block */
bh2 = sb_bread(sb, eblock->extents[ei].ee_start + bi); /* re-read dir block */
inode = simplefs_iget(sb, dblock->files[fi].inode);    /* needed only the ino_t */
```

`__file_lookup` already had the inode number. Two extra `sb_getblk` round-trips per successful lookup, on the exact path the PR cover letter is selling as faster. The buffer cache makes the I/O cheap but not free — and this is the hottest path in the filesystem.

**Fix:** have `__file_lookup` return the inode number via an out-pointer; drop the second `sb_bread` pair entirely. `simplefs_lookup` then collapses to roughly:
```c
err = __file_lookup(dir, dentry, &ino);
switch (err) {
case 0:        inode = simplefs_iget(sb, ino); break;
case -ENOENT:  inode = NULL; break;
default:       return ERR_PTR(err);
}
return d_splice_alias(inode, dentry);
```

RoyWFHuang replied:
The same as https://github.com/sysprog21/simplefs/pull/79#discussion_r3033641198

refactoring  ```__file_lookup``` involves a larger scope and will be handled in a dedicated PR.

jserv commented on inode.c:
```diff
@@ -753,186 +856,233 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
+
+        brelse(bh_ext);
+        bh_ext = NULL;
+        goto release_new;
```
**MEDIUM — in-place same-dir rename skips directory timestamp updates**

This branch does `goto release_new` directly (line 947), bypassing the ctime/mtime block at lines 1024-1053. Userspace `stat` reports a stale `mtime` after a successful rename — observable inconsistency.

**Fix:** update `dest_dir`/`src_dir` timestamps before the goto, or restructure the success path so all rename branches funnel through a single timestamp-update block.

RoyWFHuang replied:
Fixed

RoyWFHuang committed:
Optimize inode cache read and update logic

Streamline the cache lookup for frequently accessed inodes.
Reduce unnecessary overhead in block allocation checks.

jserv reviewed (changes requested):
Rebase latest `master` branch.
